src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypingDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,148 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.Diagnostics.TypingStyles
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal sealed class CSharpUseImplicitTypingDiagnosticAnalyzer : CSharpTypingStyleDiagnosticAnalyzerBase
+    {
+
+        private static readonly LocalizableString s_Title =
+            new LocalizableResourceString(nameof(CSharpFeaturesResources.UseImplicitTypingDiagnosticTitle), CSharpFeaturesResources.ResourceManager, typeof(CSharpFeaturesResources));
+
+        private static readonly LocalizableString s_Message =
+            new LocalizableResourceString(nameof(CSharpFeaturesResources.UseImplicitTyping), CSharpFeaturesResources.ResourceManager, typeof(CSharpFeaturesResources));
+
+        private static readonly DiagnosticDescriptor s_descriptorUseImplicitTyping = new DiagnosticDescriptor(
+            id: IDEDiagnosticIds.UseImplicitTypingDiagnosticId,
+            title: s_Title,
+            messageFormat: s_Message,
+            category: DiagnosticCategory.Style,
+            defaultSeverity: DiagnosticSeverity.Info,
+            isEnabledByDefault: true);
+
+        public CSharpUseImplicitTypingDiagnosticAnalyzer() : base(s_descriptorUseImplicitTyping)
+        {
+
+        }
+
+        protected override bool IsStylePreferred(SyntaxNode declarationStatement, SemanticModel semanticModel, OptionSet optionSet, CancellationToken cancellationToken)
+        {
+            var stylePreferences = GetCurrentTypingStylePreferences(optionSet);
+
+            var isTypeApparent = IsTypeApparentInDeclaration(declarationStatement, semanticModel, stylePreferences, cancellationToken);
+            var isIntrinsicType = IsIntrinsicType(declarationStatement);
+
+            return stylePreferences.HasFlag(TypingStyles.VarForIntrinsic) && isIntrinsicType
+                || stylePreferences.HasFlag(TypingStyles.VarWhereApparent) && isTypeApparent
+                || stylePreferences.HasFlag(TypingStyles.VarWherePossible) && !(isIntrinsicType || isTypeApparent);
+        }
+
+        protected override bool AnalyzeVariableDeclaration(TypeSyntax typeName, SemanticModel semanticModel, OptionSet optionSet, CancellationToken cancellationToken, out TextSpan issueSpan)
+        {
+            issueSpan = default(TextSpan);
+
+            // If it is already var, return.
+            if (typeName.IsTypeInferred(semanticModel))
```

Comment:
I just took a look at the [`IsTypeInferred(...)`](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Workspaces/Extensions/TypeSyntaxExtensions.cs,88192edfd2833ed8) function and I'm not sure it's correct. It looks to me like it'll return true for a type, `var<T>`. It's esoteric, but could you check? If it does fail on that, please add a test and fix the `IsTypeInferred(...)` function.